python/ray/data/dataset.py

```diff
@@ -101,20 +101,27 @@
 
 @PublicAPI
 class Dataset(Generic[T]):
-    """Implements a distributed Arrow dataset.
+    """A Dataset is a distributed data collection for data loading and processing.
 
-    Datasets are implemented as a list of ``ObjectRef[Block]``. The block
-    also determines the unit of parallelism. The default block type is the
-    ``pyarrow.Table``. Arrow-incompatible objects are held in ``list`` blocks.
+    Datasets are implemented as a list of ``ObjectRef[Block]``, where each block
+    holds an ordered collection of items, representing a segment of the overall
+    data collection. The block can be either a ``pyarrow.Table``, ``pandas.DataFrame``
+    or Python list. The block also determines the unit of parallelism.
+
+    Datasets can be created in multiple ways: from sythetic data via range_.*() APIs,
+    from existing memory data via from_*() APIs, or from external storage systems
+    such as local disk, S3, HDFS etc. via the read_*() APIs. The (potentially processed)
+    Dataset can be saved back to external storage systems via the write_*() APIs.
+
+    Dataset supports parallel processing at scale: transformations such as
+    .map_batches(), aggregations such as .min()/.max()/.groupby(), sorting, shuffling,
+    and repartition etc.
```

Comment:
```suggestion
    :meth:`.map_batches() <ray.data.Dataset.map_batches>`, global aggregations such
    as :meth:`.min() <ray.data.Dataset.min>`/:ref:`.max() <ray.data.Dataset.max>`/
    :ref:`.mean() <ray.data.Dataset.mean>`, grouped aggregations via
    :ref:`.groupby() <ray.data.Dataset.groupby?`, and shuffling operations such as
     :ref:`.sort() <ray.data.Dataset.sort>`,
     :ref:`.random_shuffle() <ray.data.Dataset.random_shuffle>`, and
     :ref:`.repartition() <ray.data.Dataset.repartition>`.
```

Reply:
This is text interpreter specific and the source code reader may not work with it well.

Reply:
I think our highest priority is the API docs on our hosted docs website, cc @maxpumperla since I'm not entirely sure that's the policy.

Reply:
It makes it a lot harder to read in shells and other places though. Can we avoid these refs when possible so that all readers get a good experience?

Reply:
In this case (and most docstring cases), I think that we can get the best of both worlds by leveraging a [`:py:obj` default role](https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#role-py-obj) and/or using relative references.

Reply:
If we did both, it would just be this:
```suggestion
    `.map_batches()`, global aggregations such as `.min()`/`.max()`/`.mean()`,
    grouped aggregations via `.groupby()`, and shuffling operations such as`.sort()`,
	`.random_shuffle()`, and `.repartition()`.
```
which is even more readable than double-ticking ever method.

Reply:
Single back tick has been my go-to way to refer to entities (variable/func/class etc) in code when writing comment (I think Google's code style also recommends this). Easy to read. And the code search tool can identify that and make it easy to search things in codebase.